### PR TITLE
Domains: Add view for transfering a domain between sites

### PR DESCRIPTION
**Pull request by klimeryk**
Adds a view for transferring a domain between sites.
This complements our existing tool for transferring a domain between users of the same site. Now, the user should be able to move around the domain freely, without Happiness Engineer's intervention 🎉 

New entry in the Transfers section:
screen shot 2017-05-14 at 20 00 54

New view:
screen shot 2017-05-14 at 20 01 29

Confirmation after clicking on the selected site:
screen shot 2017-05-14 at 20 01 50

## Testing
~Requires `D5603-code` on the backend.~ Backend patch already merged.

Test a few scenarios:
* Moving a domain between sites you're admin on.
* Moving a domain to a site you're not an admin on (the site should not be visible on the site picker and the backend should reject such a request).
* Moving a domain you don't own (should show the `NonOwnerCard` component.

When the domain is moved, it will _not_ become the primary domain on the target site. The user can set it as the primary one using the usual means.

There should be no WHOIS changes with this move, of course, so no emails from Tucows/WWD are expected as a result.

### CODE REVIEW

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+	render() {
+		const { selectedSite, selectedDomainName } = this.props,
+			{ slug } = selectedSite;
+
+		return (
+			<Main className="transfer-to-other-site">
+				<Header
+					selectedDomainName={ selectedDomainName }
+					backHref={ paths.domainManagementTransfer( slug, selectedDomainName ) }>
+					{ this.props.translate( 'Transfer Domain To Another Site' ) }
```
Comment (a8ci18n) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **18** times:
`translate( 'Transfer Domain To Another User' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+				showConfirmationDialog: true,
+			} );
+		}
+	}
+
+	handleConfirmTransferDomain = ( closeDialog ) => {
+		const { selectedDomainName } = this.props,
+			targetSiteName = this.state.targetSite.name,
+			successMessage = this.props.translate(
+				'%(selectedDomainName)s has been transferred to site: %(targetSiteName)s',
```
Comment (a8ci18n) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '%(selectedDomainName)s has been transferred to %(selectedUserDisplay)s' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/upgrades/domain-management/transfer/index.jsx [hunk 3]
```diff
@@ -29,15 +29,19 @@ function Transfer( props ) {
...
-					{ ! isAutomatedTransfer && ! isDomainOnly &&
-						<VerticalNavItem path={ paths.domainManagementTransferToAnotherUser( slug, selectedDomainName ) }>
-							{ translate( 'Transfer to another user' ) }
-						</VerticalNavItem> }
+				{ ! isAutomatedTransfer && ! isDomainOnly &&
+					<VerticalNavItem path={ paths.domainManagementTransferToAnotherUser( slug, selectedDomainName ) }>
+						{ translate( 'Transfer to another user' ) }
+					</VerticalNavItem>
+				}
+				{ ! isAutomatedTransfer && ! isDomainOnly &&
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: I've mimicked the conditions for transferring a domain between users - any objections?

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 1 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Sorry, but `Site` != `User` 😞

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 2 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: It's my fault for making those strings so similar, but that's how it is :man_shrugging:

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+			</div>
+		);
+	}
+}
+
+export default connect(
+	state => ( {
+		currentUser: getCurrentUser( state ),
+		sites: getSites( state ),
+		siteBySlug: ( siteSlug ) => getSiteBySlug( state, siteSlug ),
```
Comment (mcsf) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Drive-by note: please have a look at #14024, since this is doing the same thing with different syntax. :)

Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Thanks for the heads-up! I'll rework this bit and try to avoid the antipattern then 🙇

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+			targetSiteSlug,
+			showConfirmationDialog: true,
+		} );
+	}
+
+	handleConfirmTransfer = ( targetSite, closeDialog ) => {
+		const { selectedDomainName } = this.props,
+			targetSiteName = targetSite.name,
+			successMessage = this.props.translate(
+				'%(selectedDomainName)s has been transferred to site: %(targetSiteName)s',
```
Comment (a8ci18n) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '%(selectedDomainName)s has been transferred to %(selectedUserDisplay)s' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by klimeryk** [comment]
Redux antipattern squashed ;) Ready for another review 🙇

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+			.then( () => {
+				this.setState( { disableDialogButtons: false } );
+				this.props.successNotice( successMessage, { duration: 10000, isPersistent: true } );
+				closeDialog();
+				page( paths.domainManagementList( this.props.selectedSite.slug ) );
+			}, error => {
+				this.setState( { disableDialogButtons: false } );
+				this.props.errorNotice( error.message || defaultErrorMessage );
+				closeDialog();
+			} );
```
Comment (mcsf) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: This is a precursor to my next question:

Typically, the `wpcom` call wouldn't happen in the component; instead, this would be inside an action creator. Is there a reason to inline it here? (I understand one would want to avoid the overhead of having an action.)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+		}
+	}
+
+	render() {
+		if ( ! this.isDataReady() ) {
+			return <DomainMainPlaceholder goBack={ this.goToEdit } />;
+		}
+
+		const { selectedSite, selectedDomainName } = this.props,
+			{ slug } = selectedSite;
```
Comment (mcsf) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: `const` :)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			targetSiteSlug: '',
+			showConfirmationDialog: false,
+			disableDialogButtons: false
+		};
+	}
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Nice, will do!

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+import DomainMainPlaceholder from 'my-sites/upgrades/domain-management/components/domain/main-placeholder';
+import SectionHeader from 'components/section-header';
+import TransferConfirmationDialog from './confirmation-dialog';
+import { successNotice, errorNotice } from 'state/notices/actions';
+import wp from 'lib/wp';
+
+const wpcom = wp.undocumented();
+
+class TransferToOtherSite extends React.Component {
+
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: It doesn't seem to? (or maybe I've old configuration somewhere?) But nevertheless, I'll remove it, of course.

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+			showConfirmationDialog: false,
+			disableDialogButtons: false
+		};
+	}
+
+	isDataReady() {
+		return this.props.domains.hasLoadedFromServer;
+	}
+
+	checkSiteEligibility = ( site ) => {
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: `checkSiteEligibility` seemed off to me too, but I didn't realize why. Thinking about it as a predicate makes total sense 👍

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 6 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Nice, didn't know you could chain the promises like that 👍

Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: > Typically, the wpcom call wouldn't happen in the component; instead, this would be inside an action creator. Is there a reason to inline it here? (I understand one would want to avoid the overhead of having an action.)

The reason is much simpler - I've based this component on https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/upgrades/domain-management/transfer/transfer-to-other-user/index.jsx and apparently did not bother to see if the code there is now not up to par with our current guidelines. Will update 👍

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+
+const wpcom = wp.undocumented();
+
+class TransferToOtherSite extends React.Component {
+
+	static propTypes = {
+		selectedDomainName: React.PropTypes.string.isRequired,
+		selectedSite: React.PropTypes.oneOfType( [
+			React.PropTypes.object,
+			React.PropTypes.bool
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Seems like this is an unnecessary safeguard - it should not be possible to get to this view without selecting a site first. And this action does not make sense for all sites either, so I'll just update it to expect `object`.

File: client/my-sites/upgrades/domain-management/controller.jsx [hunk 12]
```diff
@@ -312,6 +312,30 @@ export default {
...
+		if ( isAutomatedTransfer ) {
+			const siteSlug = getSelectedSiteSlug( state );
+			page.redirect( `/domains/manage/${ siteSlug }` );
+			return;
+		}
+
+		setTitle(
+			i18n.translate( 'Transfer Domain' ),
+			pageContext
+		);
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/controller.jsx]: I like the idea, but I'd rather keep it consistent with the rest of the file for now. Refactoring `Header` in a separate PR (and addressing all the cases) seems like the best way to go 👍 I've created an issue to remember about this: https://github.com/Automattic/wp-calypso/issues/14295

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 6 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Sorry for the late reply, @mcsf, but I had to brush up on my Redux... And I'm still puzzled why you think an action creator is needed here? I can't foresee another component using this action, so the DRY principle does not apply. Unless I'm missing something, in this case it seems it does not make sense to use one. I'd understand if it was a convention or in the guidelines, but I can't find anything about this being a rule.

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+			targetSiteSlug,
+			showConfirmationDialog: true,
+		} );
+	}
+
+	handleConfirmTransfer = ( targetSite, closeDialog ) => {
+		const { selectedDomainName } = this.props;
+		const targetSiteName = targetSite.name;
+		const successMessage = this.props.translate(
+				'%(selectedDomainName)s has been transferred to site: %(targetSiteName)s',
```
Comment (a8ci18n) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Hi! I've found a possible matching string that has already been translated **17** times:
`translate( '%(selectedDomainName)s has been transferred to %(selectedUserDisplay)s' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 6 above)
Comment (mcsf) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: It was suggested more out of a wish to separate concerns and move effectful logic out of the components. But, again, a suggestion, and for the scope of this PR I would never want to block you. :)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+		wpcom.transferToSite( this.props.selectedSite.ID, this.props.selectedDomainName, targetSite.ID )
+			.then(
+				() => {
+					this.props.successNotice( successMessage, { duration: 10000, isPersistent: true } );
+					page( paths.domainManagementList( this.props.selectedSite.slug ) );
+				}, ( error ) => {
+					this.props.errorNotice( error.message || defaultErrorMessage );
+				} )
+			.then( () => {
+				this.setState( { disableDialogButtons: false } );
```
Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: If you hit the `success` above, this component is unmounted, so you shouldn't call `setState` on it. If I'm reading it correctly, the second `then` code should go in the `error` handler.

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+		const defaultErrorMessage = this.props.translate(
+				'Failed to transfer %(selectedDomainName)s, please try again or contact support.', {
+					args: { selectedDomainName } } );
+
+		this.setState( { disableDialogButtons: true } );
+		wpcom.transferToSite( this.props.selectedSite.ID, this.props.selectedDomainName, targetSite.ID )
+			.then(
+				() => {
+					this.props.successNotice( successMessage, { duration: 10000, isPersistent: true } );
+					page( paths.domainManagementList( this.props.selectedSite.slug ) );
```
Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: What do you think about defaulting to the `targetSite` instead? That way you automatically refetch `/domains`, otherwise I'd suggest refetching them, as the domain is still shown in the list after a successful transfer.

File: client/my-sites/upgrades/domain-management/transfer/index.jsx [hunk 16]
```diff
@@ -29,15 +29,19 @@ function Transfer( props ) {
...
-					{ ! isAutomatedTransfer && ! isDomainOnly &&
-						<VerticalNavItem path={ paths.domainManagementTransferToAnotherUser( slug, selectedDomainName ) }>
-							{ translate( 'Transfer to another user' ) }
-						</VerticalNavItem> }
+				{ ! isAutomatedTransfer && ! isDomainOnly &&
+					<VerticalNavItem path={ paths.domainManagementTransferToAnotherUser( slug, selectedDomainName ) }>
+						{ translate( 'Transfer to another user' ) }
+					</VerticalNavItem>
+				}
+				{ ! isAutomatedTransfer &&
```
Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: Mmmm, atm the `Transfer To Other Site` path isn't whitelisted for domain only. Shouldn't we disallow this?

Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: Also, if we allow, we should delete the domain-only site on transfer. :)

Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: ^ AFAIK, that should be done automagically? Or is that only on cancelling the domain?

And, yes, it should be whitelisted, will fix - nice catch 👍

Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: Only on cancel I think. I even think you disabled the transfer at some point, no? :laughing:

Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: Actually yeah, it should work. :)

Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/index.jsx]: 😌

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 15 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: I actually prefer the source site - for it looks like this: you transfer the domain to another site, you get redirected to Domain Management for the loosing site, that triggers the domains to be refetched and you see the domain disappearing ;) Does that not look like that for you?

Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Mmm, domains weren't refetched for me, and the domain was stuck in there until I switched sites. :(

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 14 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Good catch 👍 Weird, though, seemed to have been working, but probably due to a race condition. You're right, though, that closing the dialog should only be needed in the `error` case. Will test and adjust.

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 15 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: That's weird - I've tested it quite a few times and the behaviour was always as described above 🤔

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 14 above)
Comment (mcsf) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: My bad. 😅

Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: No worries, it was a valid comment and I should have spotted the issue too :)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 6 above)
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Ah, I see now - I'll definitely keep this in mind for future PRs 👍

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx (hunk 15 above)
Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Doing another check on it...

Comment (aidvu) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Ok, I see the problem. If you open a site and it starts fetching the domains but doesn't finish before you transfer, it doesn't run the `/domains` request again and returns the old ones, so it gets stuck.

Probably a super edge case as I have 10 domains per site, but would be nice to pick it up at some point later. :)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+	isDataReady() {
+		return this.props.domains.hasLoadedFromServer;
+	}
+
+	isSiteEligible = ( site ) => {
+		return site.capabilities.manage_options &&
+			! site.jetpack &&
+			! get( site, 'options.is_domain_only', false ) &&
+			! ( this.props.domainsWithPlansOnly && get( site, 'plan.product_slug' ) === PLAN_FREE ) &&
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: If the user falls under DWPO rules and the target site has `Free` plan - deny it. It's not the best UX, but might create less confusion than when the user renews the domain next year and will be force-fed a paid plan then ;)

File: client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	renderSection() {
+		const { currentUserCanManage } = getSelectedDomain( this.props );
+		if ( ! currentUserCanManage ) {
+			return <NonOwnerCard { ...omit( this.props, [ 'children' ] ) } />;
+		}
+
+		const { selectedDomainName: domainName, domainsWithPlansOnly, translate } = this.props;
+		let message;
+		if ( domainsWithPlansOnly ) {
+			message = translate( 'Please choose a site with a paid plan ' +
```
Comment (klimeryk) [inline client/my-sites/upgrades/domain-management/transfer/transfer-to-other-site/index.jsx]: Added the `site with a paid plan` part when we're dealing with a DWPO user to hint why some sites are hidden.

**Comment by aidvu** [comment]
Great catch! :shipit:
